### PR TITLE
security: guard plugin execution timeout against NaN env var

### PR DESCRIPTION
**Pull request by alex-budanov**
## Summary

`PLUGIN_EXECUTION_TIMEOUT_MS` was parsed via `parseInt(env, 10)` behind a truthy check on the env string. A non-numeric value (`"15s"`, whitespace, anything not starting with a digit) produced `NaN`, which propagated through `options?.timeout ?? envTimeout` — `NaN` is not nullish, so `??` doesn't fall back — and ended up as `script.run({ timeout: NaN })`. That's undefined behaviour in isolated-vm and likely disables the wall-clock kill that bounds plugin execution. The 15-second timeout is one of Layer 2's two enforcement mechanisms (the other is the 128 MB memory cap), so a misconfigured deployment could silently lose half its runtime defence.

## Fix

Replace `parseInt` + truthy guard with `Number()` + `Number.isFinite() && > 0`. Rejects NaN, infinities, zero, and negatives, falling back to the 15-second default whenever the env var isn't a positive finite number.

```diff
-    const envTimeout = process.env.PLUGIN_EXECUTION_TIMEOUT_MS
-      ? parseInt(process.env.PLUGIN_EXECUTION_TIMEOUT_MS, 10)
-      : 15000;
+    const envParsed = Number(process.env.PLUGIN_EXECUTION_TIMEOUT_MS);
+    const envTimeout = Number.isFinite(envParsed) && envParsed > 0 ? envParsed : 15000;
```

## Test plan

- [x] 10 new constructor-config test cases in `tests/integrations/plugin-executor-config.test.ts`: unset / empty string / whitespace / non-numeric / zero / negative / Infinity / `options.timeout` precedence (with valid and malformed env)
- [x] Full unit suite: 2402/2402 passing
- [x] No other call sites — `grep -rn parseInt(process.env packages/backend/src/integrations/security` returned only the fixed line

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Hardened validation for plugin execution timeout and memory-limit inputs: non-finite, zero, negative, fractional or malformed values are rejected and reliably fall back to safe defaults; validated constructor options take precedence when appropriate.

* **Tests**
  * Added tests covering timeout and memory-limit defaults, env-var and option precedence, and handling of malformed or edge-case values.

[Review Change Stack](https://app.coderabbit.ai/change-stack/apex-bridge/bugspotter/pull/127)

### CODE REVIEW

File: packages/backend/src/integrations/security/plugin-executor.ts [hunk 1]
```diff
@@ -74,9 +74,12 @@ export class SecurePluginExecutor {
...
     this.analyzer = new CodeSecurityAnalyzer();
-    const envTimeout = process.env.PLUGIN_EXECUTION_TIMEOUT_MS
-      ? parseInt(process.env.PLUGIN_EXECUTION_TIMEOUT_MS, 10)
-      : 15000; // Default 15 seconds (allows 10s HTTP timeout + overhead)
+    // Default 15 s (allows 10 s HTTP timeout + overhead). Reject NaN, ≤ 0, and
+    // non-finite values from a misconfigured env var — a non-numeric string
+    // like "15s" would otherwise produce NaN and silently disable the
+    // wall-clock kill that bounds plugin execution.
+    const envParsed = Number(process.env.PLUGIN_EXECUTION_TIMEOUT_MS);
+    const envTimeout = Number.isFinite(envParsed) && envParsed > 0 ? envParsed : 15000;
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/integrations/security/plugin-executor.ts]: Fixed in 4e1b794. Extracted `isPositiveFinite()` and applied it to both `options.timeout` and `options.memoryLimit`, mirroring the env-var guard. The class-name contract holds for all numeric inputs now, not just external ones.

File: packages/backend/tests/integrations/plugin-executor-config.test.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    process.env[ENV_KEY] = '8000';
+    const executor = new SecurePluginExecutor({ timeout: 3000 });
+    expect((executor as any).defaultTimeout).toBe(3000);
+  });
+
+  it('options.timeout takes precedence even when env var is malformed', () => {
+    process.env[ENV_KEY] = 'not-a-number';
+    const executor = new SecurePluginExecutor({ timeout: 3000 });
+    expect((executor as any).defaultTimeout).toBe(3000);
+  });
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/tests/integrations/plugin-executor-config.test.ts:91]: Added in 4e1b794: 4 cases each for invalid `options.timeout` and `options.memoryLimit` (NaN / 0 / negative / Infinity), bringing the file to 20 tests.

File: packages/backend/src/integrations/security/plugin-executor.ts [hunk 3]
```diff
@@ -74,9 +74,12 @@ export class SecurePluginExecutor {
...
     this.defaultTimeout = options?.timeout ?? envTimeout;
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/integrations/security/plugin-executor.ts]: Fixed in 4e1b794. Took the same shape you suggested but factored the check into an `isPositiveFinite()` helper since the guard is now used on three numeric inputs (env, `options.timeout`, `options.memoryLimit`). Tests cover NaN / 0 / negative / Infinity for both options.

**Comment by alex-budanov** [comment]
@claude review PR

File: packages/backend/src/integrations/security/plugin-executor.ts [hunk 4]
```diff
@@ -13,6 +13,10 @@ import { ERROR_CODES } from '../plugin-utils/errors.js';
 
 const logger = getLogger();
 
+function isPositiveFinite(value: unknown): value is number {
+  return typeof value === 'number' && Number.isFinite(value) && value > 0;
+}
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/integrations/security/plugin-executor.ts]: Adopted in bc6d863. `isPositiveFinite` now uses `Number.isSafeInteger`; subtle behaviour change is that env=\"15000.5\" used to silently truncate to 15000 via `parseInt` and now falls back to the default — that's an improvement (the original input was a typo). Added two tests: `options.timeout = 1.5` and env=`15000.5` both fall back.

File: packages/backend/src/integrations/security/plugin-executor.ts [hunk 5]
```diff
@@ -74,11 +78,15 @@ export class SecurePluginExecutor {
...
-    this.defaultMemoryLimit = options?.memoryLimit ?? 128; // 128 MB
+    // Validate every numeric source — env var AND constructor options. NaN, ≤ 0,
+    // and non-finite values must all fall back to the safe defaults; otherwise a
+    // misconfigured env var ("15s" → NaN) or a typed-any caller passing
+    // `options.timeout = NaN` would silently disable the wall-clock kill /
+    // memory cap that bound plugin execution.
+    const envParsed = Number(process.env.PLUGIN_EXECUTION_TIMEOUT_MS);
+    const envTimeout = isPositiveFinite(envParsed) ? envParsed : 15000;
+    this.defaultTimeout = isPositiveFinite(options?.timeout) ? options!.timeout! : envTimeout;
+    this.defaultMemoryLimit = isPositiveFinite(options?.memoryLimit) ? options!.memoryLimit! : 128;
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/integrations/security/plugin-executor.ts]: Adopted in bc6d863. `const { timeout, memoryLimit } = options ?? {};` — TypeScript narrows the locals naturally after the `isPositiveFinite` guard, so the `options!.timeout!` non-null-assertion dance is gone. Cleaner.

File: packages/backend/src/integrations/security/plugin-executor.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/integrations/security/plugin-executor.ts]: Added in bc6d863: `options.timeout = NaN` with env=\"8000\" → `defaultTimeout = 8000`. Closes the fallback-chain coverage gap — prior tests only asserted the default-path, missing whether env still wins over the hard-coded default when options is invalid.

File: packages/backend/src/integrations/security/plugin-executor.ts [hunk 6]
```diff
@@ -13,6 +13,13 @@ import { ERROR_CODES } from '../plugin-utils/errors.js';
 
 const logger = getLogger();
 
+function isPositiveFinite(value: unknown): value is number {
+  // Timeouts (ms) and memory limits (MB) are integer-domain config values —
+  // accepting fractional inputs like `1.5` is almost always a typo. Reject
+  // anything that isn't a positive safe integer; fall back to the default.
+  return typeof value === 'number' && Number.isSafeInteger(value) && value > 0;
+}
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/integrations/security/plugin-executor.ts]: Adopted in d31121b. `isPositiveFinite` → `isPositiveSafeInteger` (the name now matches the body's `Number.isSafeInteger` check). Default constants `DEFAULT_TIMEOUT_MS` and `DEFAULT_MEMORY_LIMIT_MB` extracted with one-line comments documenting the chosen values.

File: packages/backend/src/integrations/security/plugin-executor.ts [hunk 7]
```diff
@@ -74,11 +81,16 @@ export class SecurePluginExecutor {
...
+    // Validate every numeric source — env var AND constructor options. NaN, ≤ 0,
+    // and non-finite values must all fall back to the safe defaults; otherwise a
+    // misconfigured env var ("15s" → NaN) or a typed-any caller passing
+    // `options.timeout = NaN` would silently disable the wall-clock kill /
+    // memory cap that bound plugin execution.
+    const envParsed = Number(process.env.PLUGIN_EXECUTION_TIMEOUT_MS);
+    const envTimeout = isPositiveFinite(envParsed) ? envParsed : 15000;
+    const { timeout, memoryLimit } = options ?? {};
+    this.defaultTimeout = isPositiveFinite(timeout) ? timeout : envTimeout;
+    this.defaultMemoryLimit = isPositiveFinite(memoryLimit) ? memoryLimit : 128;
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/integrations/security/plugin-executor.ts]: Applied as-suggested in d31121b — constructor now uses `isPositiveSafeInteger` + `DEFAULT_TIMEOUT_MS` / `DEFAULT_MEMORY_LIMIT_MB` consistently.

**Comment by alex-budanov** [comment]
@claude review this PR

**Comment by alex-budanov** [comment]
@coderabbitai review
